prometheus/src/main/java/com/uber/m3/tally/prometheus/PrometheusReporter.java

```diff
@@ -0,0 +1,442 @@
+ *                                                   .registry(registry)
+ *                                                   .build();
+ *   Scope scope = new RootScopeBuilder().reporter(reporter)
+ *                                       .reportEvery(Duration.ofSeconds(1));
+ *   Counter counter = scope.tagged(Collections.singletonMap("foo", "bar"))
+ *                          .counter("counter");
+ *   while (true) {
+ *      counter.inc(1);
+ *      Thread.sleep(500);
+ *   }
+ * }
+ * </pre>
+ */
+public class PrometheusReporter implements StatsReporter {
+
+    static final String METRIC_ID_KEY_VALUE = "1";
+    private static final String PREFIX_SPLITTER = "+";
+    private static final String KEY_PAIR_SPLITTER = ",";
+    private static final String KEY_NAME_SPLITTER = "=";
+    private static final String KEY_PAIR_TEMPLATE = "%s" + KEY_NAME_SPLITTER + "%s";
+    private static final TimerType DEFAULT_TIMER_TYPE = TimerType.SUMMARY;
+    private static final String[] EMPTY_STRING_ARRAY = new String[0];
+
+    private final CollectorRegistry registry;
+    private final TimerType timerType;
+    private final Map<Double, Double> defaultQuantiles;
+    private final double[] defaultBuckets;
+    private final int ageBuckets;
+    private final long maxAgeSeconds;
+    private final ConcurrentMap<String, Counter> registeredCounters;
+    private final ConcurrentMap<String, Gauge> registeredGauges;
+    private final ConcurrentMap<String, Histogram> registeredHistograms;
+    private final ConcurrentMap<String, Summary> registeredSummaries;
+
+    private PrometheusReporter(
+            Map<Double, Double> defaultQuantiles,
+            double[] defaultBuckets,
+            TimerType defaultTimerType,
+            CollectorRegistry registry,
+            int ageBuckets,
+            long maxAgeSeconds
+    ) {
+        this.registry = registry;
+        this.timerType = defaultTimerType;
+        this.defaultBuckets = defaultBuckets;
+        this.defaultQuantiles = defaultQuantiles;
+        this.ageBuckets = ageBuckets;
+        this.maxAgeSeconds = maxAgeSeconds;
+        this.registeredCounters = new ConcurrentHashMap<>();
+        this.registeredGauges = new ConcurrentHashMap<>();
+        this.registeredSummaries = new ConcurrentHashMap<>();
+        this.registeredHistograms = new ConcurrentHashMap<>();
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @Override
+    public void reportCounter(String name, Map<String, String> tags, long value) {
+        final Map<String, String> ttags = (tags == null) ? Collections.emptyMap() : tags;
+        String collectorName = canonicalMetricId(name, ttags.keySet());
+        registeredCounters.computeIfAbsent(collectorName, key -> Counter.build()
+                .name(name)
+                .help(String.format("%s counter", name))
+                .labelNames(collectionToStringArray(ttags.keySet()))
```

Comment:
Re-evaluate all hot-path methods in terms of memory churn: reduce allocations in the hot-path to a minimum and avoid copying static arrays at all costs. Instead pre-initialize static arrays (like buckets, tags, etc) during initialization and re-use it across the board.

Reply:
Could you please be more specific? Which hot-path are you referring to? Static buckets for timers are already preinitialized. Tags are not static and passed as arguments to `report<Something>` methods. All array allocations happen only during the initialization of Prometheus collectors, calculation of canonical metric IDs and passing tags values to Prometheues collectors' `collect` methods. We cannot do much about it unless rewriting the collectors implementations ourself. 
If there is something particular in the implementation, which you think can be improved, please be more specific and I'd love to improve it. As a follow up I'm planning to add benchmark tests for all the implementations of Reporter interface, which would give us more clear picture about performance.

Reply:
+1 to both points--yes we should optimize; doing it as a followup with benchmarks makes reasonable sense to me.

@alexeykudinkin on the benchmarking point--any suggestions on good approaches for profiling allocations/garbage creation? 

> Instead pre-initialize static arrays (like buckets, tags, etc) during initialization and re-use it across the board.

Good call; we could probably do this kind of thing even in absence of good data (though of course, it'd be hard to tell the impact...).